### PR TITLE
Convert ABDM into Plug - Part 1/3

### PR DESCRIPTION
**Pull request by khavinshankar**
This pull request represents the first phase of converting the ABDM system into a plug. The primary focus of this PR is to refactor the relationship between the ABHA number and the patient model, alongside improvements to signal handling and routing. Detailed changes include:

- **Model Relationship Reversal:** Changed the relationship dynamics by moving the abha_number field from the patient model to the abha_number model. This involves making patient a field within the abha_number model to better align with our new architecture.

- **Signal Creation for Care Context:** Implemented a new signal that triggers the creation of a care context whenever a consultation is saved. This enhances the functionality and integration of care context within our system.

- **ABHA Number API Enhancements:** Introduced a new abha_number viewset, serializer, and routes. These additions are designed to streamline the handling and accessibility of ABHA numbers within our system.

This PR lays the groundwork for the subsequent phase, which will further integrate and optimize the ABDM plug system.

### CODE REVIEW

File: care/abdm/api/viewsets/healthid.py [hunk 1]
```diff
@@ -293,7 +293,7 @@ def get_abha_card(self, request):
         if not patient:
             raise ValidationError({"patient": "Not Found"})
 
-        if not patient.abha_number:
+        if hasattr(patient, "abha_number") is False or patient.abha_number is None:
```
Comment (sainak) [inline care/abdm/api/viewsets/healthid.py]:
```suggestion
        if getattr(patient, "abha_number", None) is None:
```

File: care/abdm/api/viewsets/healthid.py [hunk 2]
```diff
@@ -460,6 +460,15 @@ def add_care_context(self, request, *args, **kwargs):
                 status=status.HTTP_404_NOT_FOUND,
             )
 
+        if (
+            hasattr(consultation.patient, "abha_number") is False
+            or consultation.patient.abha_number is None
+        ):
```
Comment (sainak) [inline care/abdm/api/viewsets/healthid.py]:
```suggestion
        if getattr(consultation.patient, "abha_number", None) is None:
```

File: care/abdm/api/viewsets/healthid.py [hunk 3]
```diff
@@ -507,10 +516,16 @@ def patient_sms_notify(self, request, *args, **kwargs):
 
         if not patient:
             return Response(
-                {"consultation": "No matching records found"},
+                {"patient": "No matching records found"},
                 status=status.HTTP_404_NOT_FOUND,
             )
 
+        if hasattr(patient, "abha_number") is False or patient.abha_number is None:
```
Comment (sainak) [inline care/abdm/api/viewsets/healthid.py]:
```suggestion
        if getattr(patient, "abha_number", None) is None:
```

File: care/abdm/migrations/0013_abhanumber_patient.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+class Migration(migrations.Migration):
+    def reverse_patient_abhanumber_relation(apps, schema_editor):
+        Patient = apps.get_model("facility", "PatientRegistration")
+        AbhaNumber = apps.get_model("abdm", "AbhaNumber")
+
+        for patient in Patient.objects.filter(abha_number__isnull=False):
+            abha_number = AbhaNumber.objects.get(id=patient.abha_number.id)
+            abha_number.patient = patient
+            abha_number.save()
```
Comment (sainak) [inline care/abdm/migrations/0013_abhanumber_patient.py]: use bulk update

File: care/abdm/receivers/consultation.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+from care.abdm.utils.api_call import AbdmGateway
+from care.facility.models import PatientConsultation
+
+
+@receiver(post_save, sender=PatientConsultation)
+def create_care_context(sender, instance, created, **kwargs):
+    patient = instance.patient
+
+    if created and hasattr(patient, "abha_number") and patient.abha_number is not None:
```
Comment (sainak) [inline care/abdm/receivers/consultation.py]:
```suggestion
    if created and getattr(patient, "abha_number", None) is not None:
```

**Review by rithviknishad — APPROVED** [review]

**Review by sainak — APPROVED** [review]

**Comment by nihal467** [comment]
@vigneshhari conditionally approving it for review alone by @sainak @rithviknishad & @vigneshhari, since ABDM fetched patient record takes some time to reflect in the platform (pending test)

File: care/abdm/api/viewsets/consent.py [hunk 6]
```diff
@@ -71,19 +69,9 @@ def create(self, request):
...
-                f"Error: ConsentViewSet::create failed to notify (consent_requests__init). Reason: {e}",
-                exc_info=True,
-            )
-            return Response(
-                {"detail": "Failed to initialize consent request"},
-                status=status.HTTP_400_BAD_REQUEST,
-            )
+        response = Gateway().consent_requests__init(consent)
+        if response.status_code != 202:
+            return Response(response.json(), status=response.status_code)
```
Comment (sainak) [inline care/abdm/api/viewsets/consent.py:74]: we had issues with abdm not actually responding with josn, we need to catch jsondecodeerror ourselves

**Review by sainak — CHANGES_REQUESTED** [review]

**Review by rithviknishad — APPROVED** [review]

**Review by sainak — APPROVED** [review]

**Comment by nihal467** [comment]
LGTM

**Review by rithviknishad — APPROVED** [review]

**Review by shivankacker — APPROVED** [review]
LGTM, merge conflicts need to be resolved

**Review by sainak — APPROVED** [review]

File: care/abdm/api/viewsets/abha_number.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+class AbhaNumberViewSet(
+    GenericViewSet,
+    RetrieveModelMixin,
+):
+    serializer_class = AbhaNumberSerializer
+    model = AbhaNumber
+    queryset = AbhaNumber.objects.all()
+    permission_classes = (IsAuthenticated,)
+
+    def get_queryset(self):
```
Comment (vigneshhari) [inline care/abdm/api/viewsets/abha_number.py]: Can we rewrite the method to get the object first and then check for permission, The current query might be expensive to execute.

**Review by vigneshhari — APPROVED** [review]
I have made a request for optimisation, resolve the merge conflicts as well.

**Review by sainak — APPROVED** [review]

**Pull request merged**